integration/e2e_test.go

```diff
@@ -0,0 +1,230 @@
+// +build end2end,docker
+
+package integration
+
+import (
+	"bytes"
+	"context"
+	"fmt"
+	"io/ioutil"
+	"log"
+	"net/http"
+	"os"
+	"os/exec"
+	"runtime"
+	"syscall"
+	"testing"
+	"time"
+
+	gcpcachev2 "github.com/envoyproxy/go-control-plane/pkg/cache/v2"
+	gcpserverv2 "github.com/envoyproxy/go-control-plane/pkg/server/v2"
+	gcpserverv3 "github.com/envoyproxy/go-control-plane/pkg/server/v3"
+	gcptest "github.com/envoyproxy/go-control-plane/pkg/test"
+	gcpresourcev2 "github.com/envoyproxy/go-control-plane/pkg/test/resource/v2"
+	gcptestv2 "github.com/envoyproxy/go-control-plane/pkg/test/v2"
+	"github.com/envoyproxy/xds-relay/internal/app/server"
+	yamlproto "github.com/envoyproxy/xds-relay/internal/pkg/util"
+	aggregationv1 "github.com/envoyproxy/xds-relay/pkg/api/aggregation/v1"
+	bootstrapv1 "github.com/envoyproxy/xds-relay/pkg/api/bootstrap/v1"
+	"github.com/onsi/gomega"
+)
+
+func TestMain(m *testing.M) {
+	// We force a 1 second sleep before running a test to let the OS close any lingering socket from previous
+	// tests.
+	time.Sleep(1 * time.Second)
+	code := m.Run()
+	os.Exit(code)
+}
+
+func TestSnapshotCacheSingleEnvoyAndXdsRelayServer(t *testing.T) {
+	if runtime.GOOS != "linux" {
```

Comment:
since there are many nuances for running this and envoy is the only out of process component, would it be better to make this a bash driven test?

Reply:
I thought about it and decided against a bash-driver test because I was envisioning a scenario where we control multiple envoy instances from the test. Imagine being able to stand up and down envoy instances in the middle of a test.

Reply:
Actually, I was looking around the solution used by gcp (namely trap the  even if made this a bash-driven test we would still suffer from the same problem (i.e. no cross-platform way of killing child processes and its descendants). The solution adopted in gcp (i.e. trapping the exit of the bash process to kill the parent envoy process) only kills the parent process. What I'm getting at is that even if we moved to bash we would still need to rely on a linux-only solution to handle this. 

Another alternative I considered was to try to make sure that ports were unique in each test, but that also brings its own set of complications and was quickly shot down by me.

Reply:
`Imagine being able to stand up and down envoy instances in the middle of a test` this is definitely a compelling scenario for avoiding a bash driver.

Reply:
@eapolinario Out of curiosity, did you test this locally on a linux system then?

Reply:
I run my g-c-p ingetration tests like this `docker run -v ~/src/go-control-plane:/go-control-plane --cpus=10 -it gcr.io/istio-testing/go-control-plane-ci:04-02-2020 /bin/bash`

Reply:
@jessicayuen , I added a script to help run these tests on a mac. Check the `run_docker.sh` file. Assuming you have the docker runtime installed on your mac (super easy to install via homebrew: `brew cask install docker`), you can run this locally by running `./scripts/run_docker.sh make e2e-tests`.